lib/active_storage/service/azure_service.rb

```diff
@@ -0,0 +1,75 @@
+require "active_support/core_ext/numeric/bytes"
+require "azure/storage"
+require "azure/storage/core/auth/shared_access_signature"
+
+class ActiveStorage::Service::AzureService < ActiveStorage::Service
+  attr_reader :client, :path, :blobs, :container, :signer
+
+  def initialize(path:, storage_account_name:, storage_access_key:, container:)
+    @client = Azure::Storage::Client.create(storage_account_name: storage_account_name, storage_access_key: storage_access_key)
+    @signer = Azure::Storage::Core::Auth::SharedAccessSignature.new(storage_account_name, storage_access_key)
+    @blobs = client.blob_client
+    @container = container
+    @path = path
+  end
+
+  def upload(key, io, checksum: nil)
+    instrument :upload, key, checksum: checksum do
+      begin
+        blobs.create_block_blob(container, key, io, content_md5: checksum)
+      rescue Azure::Core::Http::HTTPError
+        raise ActiveStorage::IntegrityError
+      end
+    end
+  end
+
+  def download(key)
+    instrument :download, key do
+      _, io = blobs.get_blob(container, key)
+      io
+    end
+  end
+
+  def delete(key)
+    instrument :delete, key do
+      blobs.delete_blob(container, key)
+    end
+  end
+
+  def exist?(key)
+    instrument :exist, key do |payload|
+      answer = blob_for(key).present?
+      payload[:exist] = answer
+      answer
+    end
+  end
+
+  def url(key, expires_in:, disposition:, filename:)
+    instrument :url, key do |payload|
+      base_url = url_for(key)
+      generated_url = signer.signed_uri(URI(base_url), false, permissions: "r",
+        expiry: format_expiry(expires_in), content_disposition: "#{disposition}; filename=\"#{filename}\"").to_s
+
+      payload[:url] = generated_url
+
+      generated_url
+    end
+  end
+
+  def url_for_direct_upload(key, expires_in:, content_type:, content_length:)
+    raise NotImplemented
+  end
+
+  private
+    def url_for(key)
+      "#{path}/#{container}/#{key}"
+    end
+
+    def blob_for(key)
+      blobs.get_blob_properties(container, key)
+    end
+
+    def format_expiry(expires_in)
+      (Time.now.utc + expires_in.to_i).iso8601
```

Comment:
Active Storage depends on Active Support, so I’d require `active_support/core_ext/integer/time` and inline this as `expiry: expires_in.seconds.from_now.iso8601`.